state/state.go

```diff
@@ -270,34 +270,51 @@ func (state *HelmState) SyncReleases(helm helmexec.Interface, additionalValues [
 }
 
 // downloadCharts will download and untar charts for Lint and Template
-func (state *HelmState) downloadCharts(helm helmexec.Interface, dir string) (map[string]string, error) {
+func (state *HelmState) downloadCharts(helm helmexec.Interface, dir string) (map[string]string, []error) {
 	temp := make(map[string]string, len(state.Releases))
+	errs := []error{}
 
-	for _, release := range state.Releases {
-		chartPath := ""
-		if pathExists(normalizeChart(state.basePath, release.Chart)) {
-			chartPath = normalizeChart(state.basePath, release.Chart)
-		} else {
-			fetchFlags := []string{}
-			if release.Version != "" {
-				chartPath = path.Join(dir, release.Name, release.Version, release.Chart)
-				fetchFlags = append(fetchFlags, "--version", release.Version)
-			} else {
-				chartPath = path.Join(dir, release.Name, "latest", release.Chart)
-			}
+	var wgFetch sync.WaitGroup
+	jobQueue := make(chan *ReleaseSpec, len(state.Releases))
+	wgFetch.Add(len(state.Releases))
+
+	for w := 1; w <= len(state.Releases); w++ {
+		go func() {
+			for release := range jobQueue {
+				state.logger.Infof("here")
+				chartPath := ""
+				if pathExists(normalizeChart(state.basePath, release.Chart)) {
+					chartPath = normalizeChart(state.basePath, release.Chart)
+				} else {
+					fetchFlags := []string{}
+					if release.Version != "" {
+						chartPath = path.Join(dir, release.Name, release.Version, release.Chart)
+						fetchFlags = append(fetchFlags, "--version", release.Version)
+					} else {
+						chartPath = path.Join(dir, release.Name, "latest", release.Chart)
+					}
 
-			// only fetch chart if it is not already fetched
-			if _, err := os.Stat(chartPath); os.IsNotExist(err) {
-				fetchFlags = append(fetchFlags, "--untar", "--untardir", chartPath)
-				if err := helm.Fetch(release.Chart, fetchFlags...); err != nil {
-					return nil, err
+					// only fetch chart if it is not already fetched
+					if _, err := os.Stat(chartPath); os.IsNotExist(err) {
+						fetchFlags = append(fetchFlags, "--untar", "--untardir", chartPath)
+						if err := helm.Fetch(release.Chart, fetchFlags...); err != nil {
+							errs = append(errs, err)
+						}
+					}
+					chartPath = path.Join(chartPath, chartNameWithoutRepository(release.Chart))
 				}
+				temp[release.Name] = chartPath
+				wgFetch.Done()
 			}
-			chartPath = path.Join(chartPath, chartNameWithoutRepository(release.Chart))
-		}
-		temp[release.Name] = chartPath
+		}()
 	}
 
+	close(jobQueue)
```

Comment:
You should enqueue all the releases into `jobQueue` immediately before this `close(jobQueue)`.
Otherwise, `for release := range jobQueue {` never receives a release because `jobQueue` is already empty and closed.

Reply:
ahhhhhhhhhhhh thank you @mumoshu !